cmake/modules/extensions.cmake

```diff
@@ -1530,11 +1530,17 @@ endfunction()
 # Usage:
 #   zephyr_build_string(<out-variable>
 #                       BOARD <board>
-#                       [BOARD_IDENTIFIER <identifier>]
+#                       [QUALIFIERS <qualifiers>]
 #                       [BOARD_REVISION <revision>]
 #                       [BUILD <type>]
 #                       [MERGE [REVERSE]]
 #   )
+#   zephyr_build_string(<out-variable>
+#                       SOC
```

Comment:
`zephyr_build_string()` is self-contained so it should be provided with the `SOC` value, or the alternative, make the `BOARD` argument optional. If QUALIFIER is provided without BOARD, then we create the build string from just the qualifiers.

Also, using the `QUALIFIERS` together with SoC should be carefully considered, especially as we have a future enhancement in #69546:
> Support SiPs. Currently a SiP must be modeled as a SoC, although this is technically imprecise.
Multiple different SiPs could share same SoC, and thus could benefit from sharing SoC information.

A SiP field in beginning of board qualifiers could easily mess up the proposal here.

What I assume users really need is support for `qualifiers` overlay and conf fragments, not SoC overlays and conf fragments.

A qualifier approach would automatically support SiP qualifier in future.
Also the qualifier fragment is easier to reason why it should include board variant, than why a SoC fragment should consider board variants.

Reply:
The reason SOC cannot be provided is because CONFIG_SOC (or SOC) is not available until kconfig cmake code is loaded, which happens after this point, so the SOC has to be extracted from the board target name directly

Reply:
> The reason SOC cannot be provided is because CONFIG_SOC (or SOC) is not available until kconfig cmake code is loaded,

Which is exactly why I wrote:
> > or the alternative, make the BOARD argument optional. If QUALIFIER is provided without BOARD, then we create the build string from just the qualifiers.

So in following cases:

the call:
```
zephyr_build_string(<out> BOARD plank)
```
returns `plank`

the call:
```
zephyr_build_string(<out> BOARD plank BOARD_QUALIFIERS /foo/bar)
```
returns `plank_foo_bar`

and finally the call:
```
zephyr_build_string(<out> BOARD_QUALIFIERS /foo/bar)
```
returns `foo_bar`

The latter is what you're actually extending the function to support in this new feature.
But removing the SOC flag also removes any assumption as to whether the `BOARD_QUALIFIERS` contains a SoC, a SiP, or something else.

There is no need for the `SOC` flag in this function.

Reply:
This is not what it does though, the included help text shows what it returns:
```
# calling
#   zephyr_build_string(build_string SOC BOARD_QUALIFIERS /soc/bar/foo MERGE)
# will return a list of the following strings
# `soc_bar_foo;soc` in `build_string` parameter.
```

Reply:
> This is not what it does though, the included help text shows what it returns:

I can see what it returns, but please note that for the SOC comment, which I'm raising in the thread, then
1) the MERGE behavior has nothing to do with having a SOC or not parameter, thus irrelevant for the comment of removing the SOC option
2) The MERGE feature is another topic commented elsewhere.